File: solutions/cpp/log-levels/log_levels.cpp
#include <string>

using namespace std;

namespace log_line
{
    // Get the message from a log line
    string message(string line)
    {
        int offset = line.find("]: ") + 3;
        return line.substr(offset);
    }

    // Get the log level from a log line
    string log_level(string line)
    {
        int offset = line.find("]: ") - 1;
        return line.substr(1, offset);
    }

    // Reformat a log line
    string reformat(string line)
    {
        string msg = message(line);
        string lvl = log_level(line);
        return msg + " (" + lvl + ")";
    }

} // namespace log_line

File: solutions/java/log-levels/src/main/java/LogLevels.java
public class LogLevels {

    public static String message(String logLine) {
        String delim = "]:";
        int offset = logLine.indexOf(delim) + delim.length();
        return logLine.substring(offset).trim();
    }

    public static String logLevel(String logLine) {
        int start = logLine.indexOf("[");
        int end = logLine.indexOf("]");
        return logLine.substring(start + 1, end).toLowerCase();
    }

    public static String reformat(String logLine) {
        return message(logLine) + " (" + logLevel(logLine) + ")";
    }
}
